extras/rya.streams/query-manager/src/main/java/org/apache/rya/streams/querymanager/QueryManager.java

```diff
@@ -0,0 +1,241 @@
+        this.queryRepos = requireNonNull(queryRepos);
+        this.scheduler = requireNonNull(scheduler);
+
+        // subscribe to the repos and sources to be notified of changes.
+        queryRepos.forEach(repo -> repo.subscribe(new QueryManagerQueryChange()));
+        sources.forEach((ryaInstance, source) -> source.subscribe(new QueryManagerSourceListener()));
+
+        // create query cache.
+        queriesCache = new HashMap<>();
+    }
+
+    /**
+     * Starts running a query.
+     *
+     * @param query - The query to run.(not null)
+     */
+    private void runQuery(final StreamsQuery query) {
+        requireNonNull(query);
+        LOG.trace("Starting Query: " + query.getSparql());
+        if (!queryExecutor.isRunning()) {
+            LOG.debug("Query Executor wasn't running, starting now.");
+            queryExecutor.startAndWait();
+        }
+
+        try {
+            queryExecutor.startQuery(ryaInstanceName, query);
+        } catch (final QueryExecutorException e) {
+            LOG.error("Failed to start query.", e);
+        }
+    }
+
+    /**
+     * Stops the specified query from running.
+     *
+     * @param queryId - The ID of the query to stop running. (not null)
+     */
+    private void stopQuery(final UUID queryId) {
+        requireNonNull(queryId);
+
+        LOG.trace("Stopping query: " + queryId.toString());
+        if (!queryExecutor.isRunning()) {
+            throw new IllegalStateException("Query Executor must be started before queries can be managed.");
+        }
+
+        try {
+            queryExecutor.stopQuery(queryId);
+        } catch (final QueryExecutorException e) {
+            LOG.error("Failed to stop query.", e);
+        }
+    }
+
+    @Override
+    protected void startUp() throws Exception {
+        LOG.trace("Starting Query Manager.");
+        queryRepos.forEach(repo -> repo.startAndWait());
+        sources.forEach((ryaInstance, source) -> source.startAndWait());
+    }
+
+    @Override
+    protected void shutDown() throws Exception {
+        LOG.trace("Stopping Query Manager.");
+        queryRepos.forEach(repo -> repo.stop());
+        sources.forEach((ryaInstance, source) -> source.stop());
+    }
+
+    /**
+     * An implementation of {@link QueryChangeLogListener} for the
+     * {@link QueryManager}.
+     * <p>
+     * When notified of a {@link ChangeType} performs one of the following:
+     * <li>{@link ChangeType#CREATE}: Creates a new query using the
+     * {@link QueryExecutor} provided to the {@link QueryManager}</li>
+     * <li>{@link ChangeType#DELETE}: Deletes a running query by stopping the
+     * {@link QueryExecutor} service of the queryID in the event</li>
+     * <li>{@link ChangeType#UPDATE}: If the query is running and the update is
+     * to stop the query, stops the query. Otherwise, if the query is not
+     * running, it is removed.</li>
+     */
+    private class QueryManagerQueryChange implements QueryChangeLogListener {
```

Comment:
Seeing how we implemented this class kind of makes me think this interface sucks. We could simplify this code a lot if the listener also provided the updated copy of the StreamsQuery since that's all you really need to do your work in here. I think the listener is registered with the repository, we the repsitory is already applying this logic.

Reply:
that would make things easier, but then the repos would have to know about a query executor...we can talk more tomorrow